target_snowflake/sinks.py

```diff
@@ -1,44 +1,140 @@
+from typing import Dict, Iterable, List, Optional
+from typing import Any, Dict, List
 
-from singer_sdk.target_base import Target
-from target_snowflake.database_target.csv_sink import CSVSink
-from target_snowflake.migrator import SnowflakeSchemaMigrator
+from singer_sdk import SQLSink, SQLTarget
 
+from target_snowflake.connector import SnowflakeConnector
+from target_snowflake.stages import NamedStage
 
-class SnowflakeSink(CSVSink):
+import csv
+
+
+class SnowflakeSink(SQLSink):
     """Snowflake target sink class."""
 
+    connector_class = SnowflakeConnector
+    stage_class = NamedStage
+
     def __init__(
         self,
-        target: Target,
+        target: SQLTarget,
         stream_name: str,
         schema: Dict,
         key_properties: Optional[List[str]],
     ) -> None:
         super().__init__(target, stream_name, schema, key_properties)
+        self.prepare_sink()
+
+    @property
+    def table_name(self) -> str:
+        return self.stream_name.split("-")[-1].upper()
+
+    @property
+    def schema_name(self) -> Optional[str]:
+        """Return the target schema name."""
+        return self.config["snowflake"]["schema"].upper()
+
+    @property
+    def database_name(self) -> Optional[str]:
+        """Return the target DB name."""
+        return self.config["snowflake"]["database"].upper()
 
-        self.connection = target.connect()
-        self._has_migrated = False
-        self.migrator = SnowflakeSchemaMigrator(sink=self)
-        self.table_schema = target.table_schema
+    def prepare_sink(self) -> None:
+        self.connector.connection.execute(
+            'CREATE SCHEMA IF NOT EXISTS "{self.schema_name}"'
+        )
+        self.connector.prepare_table(
+            self.full_table_name, schema=self.schema, primary_keys=self.key_properties
+        )
+        self.stage = self.stage_class(self)
 
     @property
     def max_size(self):
         return self.config["batch_size_rows"]
 
-    def start_batch(self, context: dict) -> None:
-        # TODO: perhaps Sync should have a callback hook at the beginning of execution?
-        if not self._has_migrated:
-            self.migrator.sync_table_schema()
-            self._has_migrated = True
-        super().start_batch(context)
+    def _write_csv(self, filepath: Path, records: Iterable[dict]) -> None:
+        """Write a CSV file."""
+        with open(filepath, "wt") as fp:
+            writer = csv.writer(fp, delimiter=",")
+            for i, record in enumerate(records, start=1):
+                if i == 1:
+                    writer.writerow(record.keys())
+
+                writer.writerow(record.values())
+        self.logger.info(f"Completed writing {i} records to file...")
+
+    def _upload_csv_to_table_stage(self, file_path: str, full_table_name: str) -> None:
+        """Upload data to the table's built-in stage."""
+        self.connector.connection.execute(
+            f"PUT file://{file_path} @%{full_table_name};"
```

Comment:
I was thinking of having the target `CREATE STAGE IF NOT EXISTS` it's own internal stage, but I suppose there's nothing wrong with table stages instead.

Reply:
@rabidaudio - Yeah, I am on the same page with you. Creating the stage is okay also - but it opens a new can of worms regarding the user providing a valid fully-qualified stage name, or if the name is auto-generated, can we reasonably find a DB and SCHEMA where we have `CREATE` access. 

I felt like using table stages was a reasonable "config-less" default.